src/problems_contact_2d_autodiff.jl

```diff
@@ -237,27 +254,65 @@ function assemble!(problem::Problem{Contact}, time::Float64,
         # at this point we have calculated contact force fc and gap for all slave elements.
         # next task is to find out are they in contact or not and remove inactive nodes
 
-        #nzgap = sort(nonzeros(sparse(ForwardDiff.get_value(gap))))
-        #info("gap: $nzgap")
-
-        for (i, j) in enumerate(sort(collect(S)))
-#           if j in props.always_inactive
-#               info("special node $j always inactive")
-#               C[:,j] = la[:,j]
-#               continue
-#           end
-            n = normals[:,j]
-            t = Q'*n
-            lan = dot(n, la[:,j])
-            lat = dot(t, la[:,j])
-
-            if lan - gap[1, j] > 0
-#               info("set node $j active, normal direction = $(ForwardDiff.get_value(n)), tangent plane = $(ForwardDiff.get_value(t))")
+        state = problem.properties.contact_state_in_first_iteration
+        if problem.properties.iteration == 1
+            info("First contact iteration, initial contact state = $state")
+            if state == :AUTO
+                avg_gap = ForwardDiff.value(mean([gap[1, j] for j in S]))
+                std_gap = ForwardDiff.value(std([gap[1, j] for j in S]))
+                if (avg_gap < 1.0e-12) && (std_gap < 1.0e-12)
+                    state = :ACTIVE
+                else
+                    state = :UNKNOWN
+                end
+                info("Average weighted gap = $avg_gap, std gap = $std_gap, automatically determined contact state = $state")
+            end
+        end
+
+        is_active = Dict{Int, Bool}()
+        condition = Dict()
+
+        for j in S
+            if j in props.always_in_contact
+                is_active[j] = true
+                continue
+            end
+            lan = dot(normals[:,j], la[:,j])
+            condition[j] = ForwardDiff.value(lan - gap[1, j])
+            is_active[j] = condition[j] > 0
+        end
+
+        if problem.properties.iteration == 1 && state == :ACTIVE
+            for j in S
+                is_active[j] = true
+            end
+        end
+
+        if problem.properties.iteration == 1 && state == :INACTIVE
+            for j in S
+                is_active[j] = false
+            end
+        end
+
+        debug("Summary of nodes")
+        for j in sort(collect(keys(is_active)))
+            n = map(ForwardDiff.value, normals[:,j])
+            debug("$j, c=$(condition[j]), s=$(is_active[j]), n=$n")
+        end
```

Comment:
Yes that seems to be quite trivial solution, just to check if Logging._root.level is DEBUG or not.